collect_app/src/main/java/org/odk/collect/android/logic/HierarchyElement.java

```diff
@@ -18,97 +18,115 @@
 
+import java.io.Serializable;
 import java.util.ArrayList;
 
-public class HierarchyElement {
+public class HierarchyElement implements Serializable {
+
+    private int type;
+    private FormIndex formIndex;
+    private ArrayList<HierarchyElement> itemsAtLevel;
+    private HierarchyElement parent;
     private String primaryText = "";
     private String secondaryText = "";
     private Drawable icon;
     private int color;
-    int type;
-    FormIndex formIndex;
-    ArrayList<HierarchyElement> children;
-
-
-    public HierarchyElement(String text1, String text2, Drawable bullet, int color, int type,
-            FormIndex f) {
-        icon = bullet;
-        primaryText = text1;
-        secondaryText = text2;
-        this.color = color;
-        formIndex = f;
-        this.type = type;
-        children = new ArrayList<HierarchyElement>();
-    }
-
 
-    public String getPrimaryText() {
-        return primaryText;
+    private HierarchyElement(HierarchyElementBuilder builder) {
+        type = builder.type;
+        formIndex = builder.formIndex;
+        itemsAtLevel = builder.itemsAtLevel;
+        parent = builder.parent;
+        primaryText = builder.primaryText;
+        secondaryText = builder.secondaryText;
+        icon = builder.icon;
+        color = builder.color;
     }
 
-
-    public String getSecondaryText() {
-        return secondaryText;
+    public Drawable getIcon() {
+        return icon;
     }
 
-
-    public void setPrimaryText(String text) {
-        primaryText = text;
+    public ArrayList<HierarchyElement> getItemsAtLevel() {
+        return itemsAtLevel;
     }
 
-
-    public void setSecondaryText(String text) {
-        secondaryText = text;
+    public HierarchyElement getParent() {
+        return parent;
     }
 
-
-    public void setIcon(Drawable icon) {
-        this.icon = icon;
+    public String getPrimaryText() {
+        return primaryText;
     }
 
-
-    public Drawable getIcon() {
-        return icon;
+    public String getSecondaryText() {
+        return secondaryText;
     }
 
-
     public FormIndex getFormIndex() {
         return formIndex;
     }
 
-
     public int getType() {
         return type;
     }
 
-
-    public void setType(int newType) {
-        type = newType;
-    }
-
-
-    public ArrayList<HierarchyElement> getChildren() {
-        return children;
-    }
-
-
-    public void addChild(HierarchyElement h) {
-        children.add(h);
-    }
-
-
-    public void setChildren(ArrayList<HierarchyElement> children) {
-        this.children = children;
-    }
-
-
-    public void setColor(int color) {
-        this.color = color;
-    }
-
-
     public int getColor() {
         return color;
     }
 
+    public static class HierarchyElementBuilder {
+        private int type;
+        private FormIndex formIndex;
+        private ArrayList<HierarchyElement> itemsAtLevel;
+        private HierarchyElement parent;
+        private String primaryText = "";
```

Comment:
Hey @shobhitagarwal1612 any particular reason why you are initializing these strings with empty? Just checking because I know that there are times when you can have logic being weirdly if you don't carry out the right checks. I personally just leave my strings as null until I am ready to initialize them with the actual values but do share your thoughts :sunglasses: